File: prototype/rails/src/test/domain/auth/email_address_test.rb
require "test_helper"

module Domain
  module Auth
    class EmailAddressTest < ActiveSupport::TestCase
      test "normalize lowercases an address" do
        assert_equal "artist@example.com", EmailAddress.normalize("Artist@Example.COM")
      end

      test "normalize trims surrounding whitespace" do
        assert_equal "artist@example.com", EmailAddress.normalize("  artist@example.com\n")
      end

      test "normalize leaves an already normal address alone" do
        assert_equal "artist@example.com", EmailAddress.normalize("artist@example.com")
      end

      test "normalize turns a missing address into an empty string" do
        assert_equal "", EmailAddress.normalize(nil)
      end

      test "an address with a local part and a dotted domain is valid" do
        assert EmailAddress.valid?("artist@example.com")
      end

      test "an address without an at sign is invalid" do
        refute EmailAddress.valid?("artist.example.com")
      end

      test "an address without a dotted domain is invalid" do
        refute EmailAddress.valid?("artist@example")
      end

      test "an address carrying whitespace is invalid" do
        refute EmailAddress.valid?("artist name@example.com")
      end

      test "a blank address is invalid" do
        refute EmailAddress.valid?("   ")
      end
    end
  end
end

File: prototype/node/src/app/core/auth/email-address.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { isEmailAddress, normalizeEmail, redactedEmail } from './email-address.ts'

test('normalizeEmail lowercases', () => {
  assert.equal(normalizeEmail('Artist@Example.COM'), 'artist@example.com')
})

test('normalizeEmail trims surrounding whitespace', () => {
  assert.equal(normalizeEmail('  artist@example.com\n'), 'artist@example.com')
})

test('normalizeEmail leaves an already-normal address unchanged', () => {
  assert.equal(normalizeEmail('artist@example.com'), 'artist@example.com')
})

test('normalizeEmail treats null as empty', () => {
  assert.equal(normalizeEmail(null), '')
})

test('normalizeEmail treats undefined as empty', () => {
  assert.equal(normalizeEmail(undefined), '')
})

test('isEmailAddress accepts a well-formed address', () => {
  assert.equal(isEmailAddress('artist@example.com'), true)
})

test('isEmailAddress rejects an address with no @', () => {
  assert.equal(isEmailAddress('artist.example.com'), false)
})

test('isEmailAddress rejects a domain with no dot', () => {
  assert.equal(isEmailAddress('artist@example'), false)
})

test('isEmailAddress rejects whitespace in the local part', () => {
  assert.equal(isEmailAddress('artist name@example.com'), false)
})

test('isEmailAddress rejects a blank string', () => {
  assert.equal(isEmailAddress('   '), false)
})

test('isEmailAddress rejects null', () => {
  assert.equal(isEmailAddress(null), false)
})

test('isEmailAddress accepts a unicode local part', () => {
  assert.equal(isEmailAddress('café@example.com'), true)
})

test('isEmailAddress accepts the shortest possible address', () => {
  assert.equal(isEmailAddress('a@b.c'), true)
})

test('isEmailAddress rejects a leading @ with no local part', () => {
  assert.equal(isEmailAddress('@example.com'), false)
})

test('isEmailAddress rejects a domain with a trailing dot and nothing after', () => {
  assert.equal(isEmailAddress('artist@example.'), false)
})

test('redactedEmail never contains the address it redacts', () => {
  assert.equal(redactedEmail('artist@example.com').includes('artist'), false)
})

test('redactedEmail is deterministic for the same address', () => {
  assert.equal(redactedEmail('artist@example.com'), redactedEmail('artist@example.com'))
})

test('redactedEmail ignores case and surrounding whitespace, like normalizeEmail', () => {
  assert.equal(redactedEmail('Artist@Example.COM'), redactedEmail('  artist@example.com  '))
})

test('redactedEmail differs for a different address', () => {
  assert.notEqual(redactedEmail('artist@example.com'), redactedEmail('other@example.com'))
})

test('redactedEmail is a 16-character hex digest', () => {
  assert.match(redactedEmail('artist@example.com'), /^[0-9a-f]{16}$/)
})
